src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java

```diff
@@ -403,19 +412,26 @@ private void register(Iterable<NoGood> noGoods, Map<Integer, NoGood> difference)
 			knownGroundingSubstitutions.get(nonGroundRule).add(substitution);
 		}
 
-		List<NoGood> generatedNoGoods = new ArrayList<>();
 		// Collect ground atoms in the body
 		ArrayList<Integer> bodyAtomsPositive = new ArrayList<>();
 		ArrayList<Integer> bodyAtomsNegative = new ArrayList<>();
 		for (Atom atom : nonGroundRule.getBodyAtomsPositive()) {
-			if (atom instanceof BuiltinAtom) {
+
+			if (atom instanceof ExternalAtom) {
+				ExternalAtom external = (ExternalAtom) atom;
+
+				if (external.hasOutput()) {
+					continue;
```

Comment:
Externals with output just bind their output terms, so they already are covered when we have obtained a ground substitution. Conversely, we will never reach a ground substitution, if an external atom has output terms but does not provide any bindings for them.

External atoms that have no output terms have no say in the process of finding a substitution, they only read the arguments they are given and evaluate them, they will never bind anything. So these need to be evaluated for every ground substitution we have obtained.

Reply:
If the external evaluates to false, then no rule with this external can fire, so we may directly stop finding a full substitution since the rule(s) will be skipped anyways. Returning `emptyList()` does the stopping.

Reply:
This is done here: https://github.com/alpha-asp/Alpha/pull/64/files#diff-3bb08e391a45e7c42bba1e6471511498R586 (with some horrible variable naming)

Reply:
Oh, I was not aware of that, sorry.